crypto/pkcs7/pkcs7.c

```diff
@@ -1086,3 +1062,347 @@ OPENSSL_END_ALLOW_DEPRECATED
+      // information that could be used in MMA.
+      if (!pkcs7_cmp_ri(ri, pcert)) {
+        break;
+      }
+      ri = NULL;
+    }
+    if (ri == NULL) {
+      OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_RECIPIENT_MATCHES_CERTIFICATE);
+      goto err;
+    }
+    // |pkcs7_decrypt_rinfo| will only return false on critical failure, not
+    // on decryption failure. Decryption check happens below, after we populate
+    // |dummy_key| with random bytes.
+    if (!pkcs7_decrypt_rinfo(&cek, ri, pkey)) {
+      goto err;
+    }
+  } else {
+    // Attempt to decrypt every recipient info. Don't exit early as
+    // countermeasure for MMA.
+    for (size_t ii = 0; ii < sk_PKCS7_RECIP_INFO_num(rsk); ii++) {
+      ri = sk_PKCS7_RECIP_INFO_value(rsk, ii);
+      uint8_t *tmp_cek;
+      // |pkcs7_decrypt_rinfo| will only return false on critical failure, not
+      // on decryption failure. Check whether |tmp_cek| is present after the
+      // call to determine if decryption succeeded.
+      if (!pkcs7_decrypt_rinfo(&tmp_cek, ri, pkey)) {
+        goto err;
+      }
+      // Set |cek| to the first successful decryption and keep going
+      if (tmp_cek && !cek) {
+        cek = tmp_cek;
+      }
+    }
+  }
+  // Clear any decryption errors to minimize behavioral difference under MMA
+  ERR_clear_error();
+
+  EVP_CIPHER_CTX *evp_ctx = NULL;
+  BIO_get_cipher_ctx(cipher_bio, &evp_ctx);
+  if (!EVP_CipherInit_ex(evp_ctx, cipher, NULL, NULL, NULL, 0)) {
+    goto err;
+  }
+  uint8_t iv[EVP_MAX_IV_LENGTH];
+  OPENSSL_memcpy(iv, enc_alg->parameter->value.octet_string->data,
+                 enc_alg->parameter->value.octet_string->length);
+  const int expected_iv_len = EVP_CIPHER_CTX_iv_length(evp_ctx);
+  if (enc_alg->parameter->value.octet_string->length != expected_iv_len) {
+    OPENSSL_PUT_ERROR(PKCS7, ERR_R_PKCS7_LIB);
+    goto err;
+  }
+  if (!EVP_CipherInit_ex(evp_ctx, NULL, NULL, NULL, iv, 0)) {
+    goto err;
+  }
+  // Get the key length from cipher context so we don't condition on |cek_len|
+  int len = EVP_CIPHER_CTX_key_length(evp_ctx);
+  if (!len) {
+    goto err;
+  }
+  // Always generate random bytes for the dummy key, regardless of |cek| decrypt
+  dummy_key = OPENSSL_malloc(len);
+  RAND_bytes(dummy_key, len);
+  // At this point, null |cek| indicates that no content encryption key was
+  // successfully decrypted. We don't want to return early due to MMA. So, swap
+  // in the dummy key and proceed. Content decryption result will be gibberish.
+  if (cek == NULL) {
+    cek = dummy_key;
+    dummy_key = NULL;
+  }
+
+  if (!EVP_CipherInit_ex(evp_ctx, NULL, NULL, cek, NULL, 0)) {
+    goto err;
+  }
+
+  OPENSSL_free(cek);
+  OPENSSL_free(dummy_key);
+  out = cipher_bio;
+
+  if (data_body && data_body->length > 0) {
+    data_bio = BIO_new_mem_buf(data_body->data, data_body->length);
+  } else {
+    data_bio = BIO_new(BIO_s_mem());
+    if (data_bio == NULL) {
+      goto err;
+    }
+    BIO_set_mem_eof_return(data_bio, 0);
```

Comment:
We should check `BIO_set_mem_eof_return` as well

Reply:
`BIO_set_mem_eof_return` returns 1 in all cases today, but i suppose that could change in the future.